proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,754 @@
+    cleanly into the existing constraint solver mechanism. Alternatively, we
+    could stick to the simple defaulting scheme mentioned in the section about
+    annotation-free arrows. But...
+
+2.  Type inference with the "simple" matchability defaulting scheme is
+    incomplete. Take following program ::
+      nested :: a b ~ c Id => b Bool
+      nested = False
+
+    Initially, the matchabilities of ``a``, ``b`` and ``c`` are all
+    instantiated with unification variables, and there are no further
+    steps. So they are all defaulted to be matchable, at which point
+    the equality can be decomposed, and we learn that
+    ``(b :: Type -> @M Type) ~ (Id :: Type -> @U Type)``. This way, ``nested`` cannot be
+    called, because no such ``b`` exists.
+
+    Instead, we could do something more clever by defaulting
+    matchabilities in dependency order, but it's not obvious if this
+    additional complexity would be worth it.
+
+3.  We could make different choices for the syntax, regarding how we annotate
+    arrows with matchabilities or the particular names around the
+    ``Matchability`` type.
+
+4.  Data constructors could be considered to have matchable types. This would
+    make promotion more unified, as promoted constructors have matchable kinds.
+    This is quite appealing, but doing so would require additional engineering
+    effort. Either we would need to introduce matchability-polymorphic
+    term-level functions, or, perhaps more realistically, eta-expand all data
+    constructor applications to demote them to unmatchable.
+
+5.  When a kind signature is given, we make the choice of not generalising the
+    matchabilities, which differs from the treatment of kind variables. Consider
+    the following program ::
+
+      type A :: Proxy a -> Type
+      type family A
+
+    The inferred kind is ``A :: forall {k} (a :: k). Proxy a -> @M Type``, so the
+    kind of the type variable ``a`` did get generalised, but the matchability of
+    the arrow didn't (note that ``A`` is nullary). An alternative option would be
+    to simply generalise these matchability variables too, and arrive at the more
+    general ``A :: forall {k} {m} (a :: k). Proxy a -> @m Type`` kind.
+
+    But we don't do this, because doing so would result in counterintuitive
+    behaviour in many common cases, in particular, type variables introduced in
+    this way could block type family reduction. Consider the following examples ::
+
+      type B :: Type -> Type
+      type family B where
+        B = Maybe
+
+      type C :: (Type -> Type) -> Type -> Type
+      type family C f where
+        C f = f
+
+   If we infer ``B :: forall {m}. Type -> @m Type``, then ``:kind! B`` is stuck! This is
+   because type variables have computational relevance in type family reduction. In other
+   words, ``B`` becomes a matchability-indexed type family, which is likely not what the user
+   intended.
```

Comment:
It might be worth mentioning here that because of the way GHC performs kind inference, it must choose to generalize matchability variables or not _before_ it ever looks at the body of type families like `B`, which is a strong motivation for the conservative defaulting behavior of matchability variables in standalone kind signatures.

Reply:
Very good point, this was indeed a very strong motivation, and it should be included in the proposal.